File: Stack/NextGreater.py
import numpy as np;
def CalculateSpan(arr):
    stack=[]
    stack.append(arr[-1])
    s=np.zeros(shape=len(arr),dtype=int)
    s[-1]=-1
    for i in range(-2,-len(arr)-1,-1):
        while(len(stack)>0 and stack[-1]<=arr[i]):
            stack.pop()
        s[i]=-1 if(len(stack)<=0) else(stack[-1])
        stack.append(arr[i])
       # print(stack)
    print(s)
CalculateSpan([5,15,10,8,6,12,9,18])

File: Stack/NextGreater.java
import java.util.*;
class NextGreater
{
    public static void main(String args[])
    {
        NextGreater(new int[]{
            5,15,10,8,6,12,9,18
        });
    }
    private static void NextGreater(int[] arr)
    {
        Stack<Integer> stack=new Stack<Integer>();
        int n=arr.length;
        int[] dp=new int[n];
        stack.push(arr[n-1]);
        dp[n-1]=-1;
        for(int i=n-2;i>=0;i--)
        {
            while(!(stack.empty()) && arr[i]>=stack.peek())
            {
                stack.pop();
            }
            if(stack.empty())
            {
                dp[i]=-1;
            }
            else
                dp[i]=stack.peek();
            stack.push(arr[i]);
        }
        for(int i=0;i<n;i++)
            System.out.print(dp[i]+" ");
    }
}